pkg/internalserver/service.go

```diff
@@ -0,0 +1,112 @@
+package internalserver
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"net/http"
+	"time"
+
+	gokitlog "github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/influx2cortex/pkg/util/log"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	DefaultListenAddress           = "0.0.0.0"
+	DefaultListenPort              = 8081
+	DefaultGracefulShutdownTimeout = 5 * time.Second
+)
+
+type ServiceConfig struct {
+	HTTPListenAddress       string
+	HTTPListenPort          int
+	GracefulShutdownTimeout time.Duration
+}
+
+func (c *ServiceConfig) RegisterFlags(flags *flag.FlagSet) {
+	flags.StringVar(&c.HTTPListenAddress, "internalserver.http-listen-address", DefaultListenAddress, "Sets the listen address for the internal http server")
+	flags.IntVar(&c.HTTPListenPort, "internalserver.http-listen-port", DefaultListenPort, "Sets listen address port for the internal http server")
+	flags.DurationVar(&c.GracefulShutdownTimeout, "internalserver.graceful-shutdown-timeout", DefaultGracefulShutdownTimeout, "Graceful shutdown period; defaults to 5 seconds")
+}
+
+type Service struct {
+	services.Service
+
+	logger gokitlog.Logger
+
+	config  ServiceConfig
+	server  *http.Server
+	errChan chan error
+}
+
+func NewService(config ServiceConfig, logger gokitlog.Logger) (*Service, error) {
+	if logger == nil {
+		return nil, errors.New("logger should not be nil")
+	}
+
+	mux := http.NewServeMux()
+
+	mux.Handle("/metrics", promhttp.Handler())
+	mux.Handle("/healthz", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		_, _ = w.Write([]byte("ok"))
+	}))
+
+	httpServer := &http.Server{
```

Comment:
does this mean the binary has two http servers (the api and the internalserver)?  That's a little weird but I can see why we might do it that way.  Though that means we need to make sure the /metrics call is going to the correct port vs the write calls.  (Not sure where promscrape is configured)

Reply:
Yep! I figured out how to do it yesterday, so I'll have a follow-up PR to get that fixed once this gets merged.